### PR TITLE
cli(asset-saver): print one devtoolsLog event per line

### PR DESCRIPTION
**Pull request by brendankenny**
It's very nice that the trace that comes out of `-GA` is saved with a trace event per line. Doing some recent work with devtoolsLogs made me really want the same printing style for those.

### CODE REVIEW

File: lighthouse-core/lib/asset-saver.js [hunk 1]
```diff
@@ -216,24 +225,24 @@ function* traceJsonGenerator(traceData) {
...
-      },
-    });
+async function saveTrace(traceData, traceFilename) {
+  const traceIter = traceJsonGenerator(traceData);
+  const writeStream = fs.createWriteStream(traceFilename);
 
-    const writeStream = fs.createWriteStream(traceFilename);
-    writeStream.on('finish', resolve);
-    writeStream.on('error', reject);
+  return pipeline(stream.Readable.from(traceIter), writeStream);
```
Comment (brendankenny) [inline lighthouse-core/lib/asset-saver.js:232]: this was able to get way nicer than the old version. Node 13+ will let us drop the `stream.Readable.from()` part (`pipeline` will take iterables directly) and node 15+ will let us drop the `promisify(stream.pipeline)` call

File: lighthouse-core/test/lib/asset-saver-test.js [hunk 2]
```diff
@@ -89,14 +89,37 @@ describe('asset-saver helper', () => {
...
+      expect(traceFileContents).toEqual(
+`{
+"traceEvents": [
+  {"args":{},"cat":"devtools.timeline","pid":1,"ts":2},
+  {"args":{},"cat":"v8","pid":1,"ts":3},
+  {"args":{"IsMainFrame":true},"cat":"v8","pid":1,"ts":5},
+  {"args":{"data":{"encodedDataLength":20,"requestId":"1.22"}},"pid":1,"ts":6}
+]}
+`);
+    });
```
Comment (brendankenny) [inline lighthouse-core/test/lib/asset-saver-test.js:113]: we didn't have any tests of the printing style, just that all the correct content was being printed in a JSON.parse()able way.

File: yarn.lock [hunk 3]
```diff
@@ -656,9 +656,9 @@
   integrity sha512-tHq6qdbT9U1IRSGf14CL0pUlULksvY9OZ+5eEgl1N7t+OA3tGvNpxJCzuKQlsNgCVwbAs670L1vcVQi8j9HjnA==
 
 "@types/node@*":
-  version "10.14.0"
-  resolved "https://registry.yarnpkg.com/@types/node/-/node-10.14.0.tgz#1c297530428c6f4e0a0a3222f5b44745669aa9f7"
-  integrity sha512-1UhSMMDix7bVdUeqtZERQQyJr3QuFoN5X5APtpIooGkumE3crPaeq7UgFeJNjGD8yCQ8od8PzRkgptR5+x327Q==
+  version "12.20.7"
```
Comment (brendankenny) [inline yarn.lock:659]: needed up to date (for Node 12) types to be able to use `stream.Readable.from()`. 

(@adamraine these outdated node types were also the thing preventing this suggestion from working: https://github.com/GoogleChrome/lighthouse/pull/12145#discussion_r598014045)

**Comment by brendankenny** [comment]
test failure is browserify not supplying `util.promisify()`, and because `asset-saver.js` is pulled into the bundle (even though it's not used in bundle code :/) it's causing an error.

It is included in browserify 17 (up from the current 16.2.3`. I'm looking at how easy the update is vs just doing a workaround.

**Comment by brendankenny** [comment]
> It is included in browserify 17 (up from the current 16.2.3`). I'm looking at how easy the update is vs just doing a workaround.

it's not bad at all

File: lighthouse-core/lib/asset-saver.js [hunk 4]
```diff
@@ -164,47 +165,55 @@ async function prepareAssets(artifacts, audits) {
...
-        eventsRemaining = EVENTS_PER_ITERATION;
-        eventsJSON = '';
+  yield '[\n';
+  if (arrayOfObjects.length > 0) {
+    const elementsIterator = arrayOfObjects[Symbol.iterator]();
+    // Emit first element manually to avoid a trailing comma.
+    const firstElement = elementsIterator.next().value;
+    yield `  ${JSON.stringify(firstElement)}`;
+
+    let elementsRemaining = ELEMENTS_PER_ITERATION;
```
Comment (paulirish) [inline lighthouse-core/lib/asset-saver.js:184]: "element" => "item"? or "value"?

File: lighthouse-core/lib/asset-saver.js [hunk 5]
```diff
@@ -164,47 +165,55 @@ async function prepareAssets(artifacts, audits) {
...
  */
-function* traceJsonGenerator(traceData) {
-  const EVENTS_PER_ITERATION = 500;
-  const keys = Object.keys(traceData);
-
-  yield '{\n';
+function* arrayOfObjectsJsonGenerator(arrayOfObjects) {
+  const ELEMENTS_PER_ITERATION = 500;
 
   // Stringify and emit trace events separately to avoid a giant string in memory.
```
Comment (paulirish) [inline lighthouse-core/lib/asset-saver.js:176]: update comment

File: lighthouse-core/lib/asset-saver.js [hunk 6]
```diff
@@ -164,47 +165,55 @@ async function prepareAssets(artifacts, audits) {
...
+}
+
+/**
+ * Generates a JSON representation of traceData line-by-line for a nicer printed
+ * version with one trace event per line.
+ * @param {LH.Trace} traceData
+ * @return {IterableIterator<string>}
+ */
+function* traceJsonGenerator(traceData) {
+  const {traceEvents, ...rest} = traceData;
```
Comment (paulirish) [inline lighthouse-core/lib/asset-saver.js:207]: def nicer than the old way. 👍

File: lighthouse-core/lib/asset-saver.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/lib/asset-saver.js:232]: add a compat TODO comment for the latter bit?

**Review by paulirish — APPROVED** [review]
coool. a few nits

File: lighthouse-core/lib/asset-saver.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/lib/asset-saver.js:176]: done

File: lighthouse-core/lib/asset-saver.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/lib/asset-saver.js:184]: done

File: lighthouse-core/lib/asset-saver.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/lib/asset-saver.js:232]: yah

**Review by chrystianrgs — APPROVED** [review]

**Review by chrystianrgs — APPROVED** [review]

**Pull request merged**